### PR TITLE
Add reflecting boundary conditions

### PR DESCRIPTION
**Pull request by pgrete**
What is says on the tin (and the docs).

Bonus: LW implosion test problem.

### CODE REVIEW

**Comment by BenWibking** [comment]
Did you disable FMA instructions with a compiler flag? I don't think the reconstruction method should matter. In Quokka, RK2+PPM gives perfect symmetry preservation, as long as you disable FMA instructions.

**Comment by pgrete** [comment]
> Did you disable FMA instructions with a compiler flag? I don't think the reconstruction method should matter. In Quokka, RK2+PPM gives perfect symmetry preservation, as long as you disable FMA instructions.

I have not tried disabling FMA.

File: docs/input.md [hunk 1]
```diff
@@ -323,3 +323,21 @@ d_log_temp_tol = 1e-8              # Tolerance in cooling table between subseque
 Finally, a more comprehensive descriptions of various conventions used for cooling
 functions (and the chosen implementation in AthenaPK) can be found in [Cooling Notes](cooling_notes.md)
 and a notebook comparing various cooling tables (and their conversion) in [cooling/cooling.ipynb](cooling/cooling.ipynb).
+
+## Boundary conditions
+
+In addition to enrolling custom boundary conditions, three general options are currently supported by default:
+
+- `periodic`: work without restrictions (implemented in upstream Parthenon)
+- `outflow`: work without restrictions (implemented in upstream Parthenon)
```
Comment (BenWibking) [inline docs/input.md:332]: Is this a constant extrapolation from the last valid cell into the ghost cells? I think it would help to say explicitly what this is.

File: inputs/lw_implode.in [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+nx3 = 1
+x3min = -0.5
+x3max = 0.5
+ix3_bc = periodic
+ox3_bc = periodic
+
+<parthenon/meshblock>
+nx1 = 512
+nx2 = 512
```
Comment (BenWibking) [inline inputs/lw_implode.in:46]: Why is there only 1 meshblock?

File: src/bvals/boundary_conditions_apk.hpp [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+namespace BoundaryFunction {
+
+using namespace parthenon::package::prelude;
+using parthenon::CoordinateDirection;
+// using parthenon::MeshBlockData;
+// using parthenon::Real;
+using parthenon::BoundaryFunction::BCSide;
+
+template <CoordinateDirection DIR, BCSide SIDE>
+void ReflectBC(std::shared_ptr<MeshBlockData<Real>> &mbd, bool coarse) {
```
Comment (BenWibking) [inline src/bvals/boundary_conditions_apk.hpp:39]: What does `coarse` mean here?

File: src/bvals/boundary_conditions_apk.hpp [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  const int offset = (2 * ref) + (INNER ? -1 : 1);
+
+  auto cons = mbd->PackVariables(std::vector<std::string>{"cons"}, coarse);
+  const bool fine = false; // no usage of fine fields in AthenaPK for now
+
+  const auto nv = IndexRange{0, cons.GetDim(4) - 1};
+  pmb->par_for_bndry(
+      "ReflectBC", nv, domain, parthenon::TopologicalElement::CC, coarse, fine,
+      KOKKOS_LAMBDA(const int &v, const int &k, const int &j, const int &i) {
+        const bool reflect = v == DIR;
```
Comment (BenWibking) [inline src/bvals/boundary_conditions_apk.hpp:80]: Why does this enum / int comparison work? If the order of the hydro components in AthenaPK were to change, would this still work?

File: src/pgen/lw_implode.cpp [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+  // Following Athena++
+  // https://github.com/PrincetonUniversity/athena/blob/1591aab84ba7055e5b356a8f069695ea451af8a0/src/pgen/lw_implode.cpp#L43
+  // to make sure the ICs are symmetric, set y0 to be in between cell centers
+  const auto x2min = pin->GetReal("parthenon/mesh", "x2min");
+  const auto x2max = pin->GetReal("parthenon/mesh", "x2max");
+  Real y0 = 0.5 * (x2max + x2min);
+  for (int j = jb.s; j <= jb.e; j++) {
+    if (coords.Xc<2>(j) > y0) {
+      // TODO(felker): check this condition for multi-meshblock setups
+      // further adjust y0 to be between cell center and lower x2 face
```
Comment (BenWibking) [inline src/pgen/lw_implode.cpp:56]: In my tests, this problem is not symmetric when there are multiple MeshBlocks. Is this why?

**Review by BenWibking — COMMENTED** [review]
Minor comments below. The major concern I have is that the symmetry does not appear to be preserved when there are multiple MeshBlocks used for the LW problem.

**Comment by BenWibking** [comment]
> > Did you disable FMA instructions with a compiler flag? I don't think the reconstruction method should matter. In Quokka, RK2+PPM gives perfect symmetry preservation, as long as you disable FMA instructions.
> 
> I have not tried disabling FMA.

For CUDA, I disabled FMA with these CMake flags:
```
if(DISABLE_FMAD)
  message(STATUS "Fused multiply-add (FMAD) is disabled for device code.")

  if(CMAKE_CUDA_COMPILER_ID STREQUAL "NVIDIA")
    add_compile_options($<$<COMPILE_LANGUAGE:CUDA>:--fmad=false>)
  endif()

  if(CMAKE_CUDA_COMPILER_ID STREQUAL "Clang")
    add_compile_options($<$<COMPILE_LANGUAGE:CUDA>:-ffp-contract=off>)
    add_compile_options($<$<COMPILE_LANGUAGE:C>:-ffp-contract=off>)
    add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-ffp-contract=off>)
  endif()

  if(CMAKE_CUDA_COMPILER_ID STREQUAL "NVHPC")
    add_compile_options($<$<COMPILE_LANGUAGE:C>:-Mnofma>)
    add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-Mnofma>)
    add_compile_options($<$<COMPILE_LANGUAGE:CUDA>:-gpu=nofma>)
  endif()
else()
  # fused MAD causes directional asymmetry -- enabled by default by nvcc and clang
  message(STATUS "Fused multiply-add (FMAD) is *enabled* for device code. Exact direction symmetry will NOT be preserved.")
endif(DISABLE_FMAD)
```

File: docs/input.md (hunk 1 above)
Comment (pgrete) [reply docs/input.md:332]: Yes, that's it. I expanded the comment.
I also added a warning when being used for subsonic flows.

File: src/bvals/boundary_conditions_apk.hpp (hunk 4 above)
Comment (pgrete) [reply src/bvals/boundary_conditions_apk.hpp:80]: Correct, it only works because of the current (fixed) ordering.
I don't see an issue right now (as no fundamental change of this is planned until the boundary comm in Parthenon is more efficient for many small variables), but to be sure that we don't miss this in the future, I added a regression test, which should fail if sth along those lines changes.

File: src/bvals/boundary_conditions_apk.hpp (hunk 3 above)
Comment (pgrete) [reply src/bvals/boundary_conditions_apk.hpp:39]: It's called/set from the Parthenon infrastructure for runs with mesh refinement and indicates when the coarse ghost cells of the given block should be filled.

File: inputs/lw_implode.in (hunk 2 above)
Comment (pgrete) [reply inputs/lw_implode.in]: Se below.

File: src/pgen/lw_implode.cpp (hunk 5 above)
Comment (pgrete) [reply src/pgen/lw_implode.cpp:56]: yes, that's my guess at the moment, too.
Given that this might turn into a rabbit hole, and, in practice, I expect that will have little impact on any  proper simulation (compared to idealized standard tests etc), I'm happy to ignore this for now.

**Comment by pgrete** [comment]
Alright, I think this is good to go now.

**Review by BenWibking — APPROVED** [review]
